### ISSUE TITLE
Add a way to track changes in side length during interaction

### ISSUE THREAD

**Issue by jessegreenberg** [phetsims/quadrilateral#25]
Discussed 11/2/21:

One of the learning goals is to know when the the quadrilateral changes and remains a parallelogram without changing side lengths.

**Comment by jessegreenberg** [comment]
We discussed this a bit in https://docs.google.com/document/d/1Rxp9QZbWDWAR8uTp7NxrEEPTRUjTnwBUdg2_xzE6Dzo/edit#.

The approach proposed there will give each Side a tolerance interval that is scaled by the side length so that we can compare quality of side lengths before and after interaciton.

**Comment by BLFiedler** [comment]
A question for meeting: When you're a parallelogram, do we compare a side against itself or against the opposite side to track side length maintenance?

Also, I am not sure of this yet, but I'd like to think some more about what impact the angleTolerance and "imperfect" parallelograms may have on the side length maintenance. If a shape is being constantly adjusted (e.g., 4 finger input or tangible input), what is the parallelogram you are trying to maintain?

**Comment by jessegreenberg** [comment]
Initial support for this was added in the above commit. I added
- The ability to "save" side lengths at the beginning of an interaction.
- `lengthsEqualToSavedProperty` to the model, which is an observable that lets us know when current lengths are equal to the saved lengths. Like `isParallelogramProperty` it is updated in step asynchronously so it it calculated only after all vertex positions have been updated for a given animation frame.
- Sides now have `lengthToleranceIntervalProperty`, which is the tolerance for that side when comparing lengths. It is a portion of the current side length.

This can be tested in this version: https://phet-dev.colorado.edu/html/quadrilateral/1.0.0-dev.12/phet/quadrilateral_en_phet.html


I added a new option to the "Success" sound design that will only play the "maintenance" sound when the shape changes, quad stays in parallelogram, AND sides remain the same length relative to the start of the interaction.
![image](https://user-images.githubusercontent.com/6396244/142021700-7612f50c-9695-49f6-b9ac-5a31e2bff247.png)

There is also a query parameter `?lengthToleranceIntervalScaleFactor` that controls the length tolerance interval. Can take values 0-1, where 0 means that lengths must be exactly equal and 1 means that the tolerance interval is equal to the side length.

**Comment by jessegreenberg** [comment]
Discussed in design meeting today: 

- With the tangible device we don't know when to save the lengths or when an interaction starts. Instead of the approach in https://github.com/phetsims/quadrilateral/issues/25#issuecomment-970428290 we will try detecting the change in length every time step/input event and comparing that to the set from the previous save. In order for this to work we also need to track the change in angles. There must be a large enough change in angles and a small enough change in lengths for it to register as a change where the sides are staying the same length. Otherwise you could slowly pull sides apart and it would register as "keeping the same length" because the change in length is small enough.

**Comment by BLFiedler** [comment]
During the meeting today I realized that we'll have this new tolerance property, but no modal feedback for folks to interact with it!

Jesse mentioned he'd probably have just swapped the maintenance sound to the "fixed-side" case, which is analogous to one to the possible "modes" we'll be implementing as folks move from parallelogram maintenance to fixed-side-parallelogram maintenance.

- [ ] Let's add a toggle or checkbox to the preferences menu to swap between these modes where the maintenance sound is associated with the fixed-side maintenance when enabled.

We'll keep it hidden for now while we decide what display elements will be in the play/control area.

**Comment by jessegreenberg** [comment]
I could see this getting tricky.

1) Since a new snapshot of quadrilateral properties are saved every time step it opens us up to performance-specific behavior, where equalities may not be the same on slower machines.
2) When do we start checking for equalities again after pulling sides apart? You could start with one set of side lengths, pull them apart, then shift shift sides again horizontally according to the learning goal and it would be "ok" with this algorithm.

**Comment by jessegreenberg** [comment]
@BLFiedler suggested that one way to improve this is to only save a snapshot when we have left the "success" case.

**Comment by jessegreenberg** [comment]
OK, I implemented a new solution for this.  Here is how it works now:

- QuadrilateralShapeModel has a `lengthsEqualToSavedProperty`. It is set every animation frame. It is true when the quad side lengths are equal to a set of previously saved values within a tolerance interval.
- If it ever toggles to false, a new set of side lengths are immediately saved as the new set.
- (If selected in Preferences) The sound for "success" when changing the shape will play when the shape changes while `lengthsEqualToSavedProperty` is true. When `lengthsEqualToSavedProperty` becomes false, the "success" sound stops playing. The new set of side lengths was immediately set so the next animation frame the `lengthsEqualToSavedProperty` will be true again. But the sound view has a time based delay to wait to start playing the success sounds again. This way we don't hear the "success" sound again until the user moves for a while while keeping the same lengths.

I added a query parameter to demonstrate this. `?showLengthAreas`. It shows how you have to move a side to hear the maintenance sound while keeping lengths the same. The stroke of the arc is the same as the length tolerance.
![image](https://user-images.githubusercontent.com/6396244/150035755-19c42c50-d497-4ed8-86ef-ce6b58f33f28.png)
If the length changes, a new set of arcs are drawn and you have to wait 1.5 seconds before hearing the "sucess" sounds again while maintaining the same lengths.

It seems to be working as expected. @BLFiedler can you please review? To test you will need to toggle the "Constant lengths for maintenance sound" in Preferences.

**Comment by BLFiedler** [comment]
First test on touchscreen felt really good with the first pass and default values! Will spend more time on this, but I think we've hit a great point to get some broader feedback.

- @jessegreenberg, can you create a dev version with all of the changes up to this point that I can pass on to the team at large?

**Comment by jessegreenberg** [comment]
Over slack I shared this version with @BLFiedler:

https://phet-dev.colorado.edu/html/quadrilateral/1.0.0-dev.15/phet/quadrilateral_en_phet.html

He mentioned he would append the query parameters of interest before sharing.

**Comment by BLFiedler** [comment]
Two requests regarding this feature - one for interviews, one for debugging

1. Query parameter for having the  checkbox "Constant lengths for maintenance sound" enabled on startup for present it in this mode for interviewees.
2. If possible, show length maintenance arcs from ?showLengthAreas when interacting with two or more vertices for debugging purposes

**Comment by jessegreenberg** [comment]
Query parameter added above.

> Show length maintenance arcs from ?showLengthAreas when interacting with two or more vertices

For any two vertices, I am not sure which arcs to show so I am just showing all of them. So it looks like this:
![image (2)](https://user-images.githubusercontent.com/6396244/151033945-cf7ad1c7-2368-4e49-bf84-e7974d0762d5.png)

Which may not be super helpful.

Alternatively, we chould just show an arc when dragging two vertices of the same side. Do you have a preference @BLFiedler?

**Comment by BLFiedler** [comment]
> I think this is what we should do. If you happen to be dragging all four vertices at the same time it will look like the above image, but if you are only dragging two vertices of the same side you will only see the arc that is relevant to you. If you are dragging vertices that are diagonal you are going to quickly exit your parallelogram state anyway.

Yeah, I agree with this. It's debugging functionality anywho, so let's not overdesign it. Let's stick to only extending the functionality to two vertices and displayed for the shared side as you suggest.

**Comment by jessegreenberg** [comment]
OK! This has been done. Anything else for this issue @BLFiedler?

**Comment by BLFiedler** [comment]
I noticed that the 4 vertex case you illustrated above is still possible in the dev17 link you shared. Did you intend to remove that? @jessegreenberg 

Otherwise this looks to be in a good position and we can close. Can create issues for tweaks or re-open if necessary.

**Comment by jessegreenberg** [comment]
I just checked in dev.17, I think it is behaving as I expected. The way it should work currently is that whatever is dragged should show arcs. If you drag a side or its two vertices you will see arcs for the vertices of that side. If you have more than one side dragged, arcs will be shown for all dragged sides.

Sorry, I think I wrote https://github.com/phetsims/quadrilateral/issues/25#issuecomment-1021493953 poorly and also misinterpreted https://github.com/phetsims/quadrilateral/issues/25#issuecomment-1021362168, so this may have been what you were expecting in that comment all along. But if not quite right it can definitely be adjusted.
